Expose API for backward execution order

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #85849
* #79056
* #87806
* __->__ #87507

In this PR:
- graph_task stores graph roots on construction so that we can later traverse through the graph 
- before the nodes are returned, they needed to be converted from raw_ptr to shared_ptr, and this should be OK because the graph is guaranteed to be alive

soulitzer committed:
Expose API for backward execution order

[ghstack-poisoned]

albanD commented on test/test_autograd.py:
```diff
@@ -3073,6 +3073,82 @@ def hook(_):
+        self.assertEqual(predicted[0], grad_fns(*actual))
+        actual = []
+
+        # When not called in a backward, empty list is returned
```
Shouldn't this be an error to be more consistent with all the other APIs getting info from the graph task?

soulitzer replied:
Yes, that sounds better

albanD commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -398,6 +398,65 @@ void add_node_to_current_graph_task_exec_info(Node* fn) {
 
+// NB: The engine itself does not use the outputs of this function.
+//
+//     WARNING: the predicted order is only correct when the following
```
Should we use the new single threaded backward to enforce this?

soulitzer replied:
Whoops was thinking of something different - yes, you're right, it's the same

soulitzer committed:
Update on "Expose API for backward execution order"


In this PR:
- adds API to retrieve execution order from current graph task
- graph_task stores graph roots on construction so that we can later traverse through the graph (currently saving shared_ptr, but maybe raw ptr is preferred).
- before the nodes are returned, they needed to be converted from raw_ptr to shared_ptr (this is OK if the graph is still alive)

TODO:
- test the case where inputs= argument matters


[ghstack-poisoned]

soulitzer committed:
Update on "Expose API for backward execution order"


In this PR:
- adds API to retrieve execution order from current graph task
- graph_task stores graph roots on construction so that we can later traverse through the graph (currently saving shared_ptr, but maybe raw ptr is preferred).
- before the nodes are returned, they needed to be converted from raw_ptr to shared_ptr (this is OK if the graph is still alive)

TODO:
- test the case where inputs= argument matters


[ghstack-poisoned]

soulitzer committed:
Update on "Expose API for backward execution order"


In this PR:
- graph_task stores graph roots on construction so that we can later traverse through the graph 
- before the nodes are returned, they needed to be converted from raw_ptr to shared_ptr, and this should be OK because the graph is guaranteed to be alive


[ghstack-poisoned]

soulitzer committed:
Update on "Expose API for backward execution order"


In this PR:
- graph_task stores graph roots on construction so that we can later traverse through the graph 
- before the nodes are returned, they needed to be converted from raw_ptr to shared_ptr, and this should be OK because the graph is guaranteed to be alive


[ghstack-poisoned]

soulitzer committed:
Update on "Expose API for backward execution order"


In this PR:
- graph_task stores graph roots on construction so that we can later traverse through the graph 
- before the nodes are returned, they needed to be converted from raw_ptr to shared_ptr, and this should be OK because the graph is guaranteed to be alive


[ghstack-poisoned]

albanD commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -398,6 +398,66 @@ void add_node_to_current_graph_task_exec_info(Node* fn) {
+    heap.push(ptr);
+  }
+
+  // Implementation notes:
```
This sounds good.
Also at this point, we clearly gave on the whole idea of ever allowing "graph surgery" where people would be allowed to manually re-stitch the graph after the fact as sequence_nr/topological_nr would be wrong.

albanD reviewed (approved):
Thanks!
We should discuss:
- How much of this we want to add to public API in autograd.graph
- Some sort of pre-backward hook for people to safely use these APIs (registered on a Tensor/grad_fn or globally).

soulitzer commented on test/test_autograd.py:
```diff
@@ -3073,6 +3073,129 @@ def hook(_):
+        self.assertEqual(names(predicted[0]), [
+            'CosBackward0', 'CosBackward0', 'SinBackward0', 'MulBackward0', ACC_GRAD
+        ])
+        # TODO: Uncomment after update to hooks behavior
```
Later in this stack, I updated the behavior of .grad wrt hooks, and that prevented me from using hooks to automatically validate the order of execution because certain nodes will be executed but still have their hooks fired.

albanD replied:
Ho ok sounds good!

soulitzer commented on test/test_autograd.py:
```diff
@@ -3073,6 +3073,129 @@ def hook(_):
+        t.register_hook(hook)
+        with torch.autograd.set_multithreading_enabled(False):
+            t.backward()
+        self.assertEqual(names(predicted[0]), [
```
Very cool!

soulitzer committed:
Update on "Expose API for backward execution order"


In this PR:
- graph_task stores graph roots on construction so that we can later traverse through the graph 
- before the nodes are returned, they needed to be converted from raw_ptr to shared_ptr, and this should be OK because the graph is guaranteed to be alive


[ghstack-poisoned]

soulitzer committed:
Update on "Expose API for backward execution order"


In this PR:
- graph_task stores graph roots on construction so that we can later traverse through the graph 
- before the nodes are returned, they needed to be converted from raw_ptr to shared_ptr, and this should be OK because the graph is guaranteed to be alive


[ghstack-poisoned]

soulitzer commented:
> How much of this we want to add to public API in autograd.graph

I think this would make sense, as the correctness is now guaranteed

> Some sort of pre-backward hook for people to safely use these APIs (registered on a Tensor/grad_fn or globally). 

This sounds quite useful for functions with multiple outputs. One way to implement this would be to use a custom function like we do for full backward hooks (users can already do this themselves too).

soulitzer commented:
@pytorchbot merge -g

soulitzer commented:
@pytorchbot merge -f

soulitzer commented:
@pytorchbot merge -f "Preexisting failures"
